File: 20230202_Thursday/20230202.go
// Leetcode: 953. Verifying an Alien Dictionary.
// https://leetcode.com/problems/verifying-an-alien-dictionary/
// = = = = = = = = = = = = = =
// Accepted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = =
// Runtime: 5 ms, faster than 22.81% of Go online submissions for Verifying an Alien Dictionary.
// Memory Usage: 3 MB, less than 5.26% of Go online submissions for Verifying an Alien Dictionary.
// 2023.02.02 Daily Challenge.
var h map[rune]rune
func isAlienSorted(words []string, order string) bool {
  h = make(map[rune]rune)
  alpha := "abcdefghijklmnopqrstuvwxyz"
  for i := 0; i < 26; i += 1 {
    key, value := order[i], alpha[i]
    h[rune(key)] = rune(value)
  }
  // New words.
  nw := make([]string, len(words))
  for i,w := range words {
    nw[i] = convert(w)
  }
  for i,j := 0, 1; j < len(nw); i,j = i+1,j+1 {
    if nw[i] > nw[j]{
      return false   
    }
  }
  return true
}

func convert(s string) string {
  answer := make([]rune,len(s))
  for i, key := range s {
    answer[i] = h[rune(key)]
  }
  return string(answer)
}

File: 20230202_Thursday/20230202.rb
# Leetcode: 953. Verifying an Alien Dictionary.
# https://leetcode.com/problems/verifying-an-alien-dictionary/
# = = = = = = = = = = = = = =
# Accepted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 98 ms, faster than 36.36% of Ruby online submissions for Verifying an Alien Dictionary.
# Memory Usage: 211 MB, less than 63.64% of Ruby online submissions for Verifying an Alien Dictionary.
# 2023.02.02 Daily Challenge.
# @param {String[]} words
# @param {String} order
# @return {Boolean}
def is_alien_sorted(words, order)
  h = order.chars.zip((?a..?z).to_a).to_h
  convert = ->s{ s.chars.map{|x| h[x]}.join}
  nw = words.map(&convert)
  (1...nw.size).each do |j|
    i = j-1
    return false if nw[i] > nw[j]
  end
  return true
end
